File: remote-mcp-server/src/ai.ts
import Anthropic from '@anthropic-ai/sdk'
import OpenAI from 'openai'
import { ANTHROPIC_API_KEY, OPENAI_API_KEY } from './config.js'

let _anthropic: Anthropic | null = null
let _openai: OpenAI | null = null

export function getAnthropic(): Anthropic {
  if (!_anthropic) {
    _anthropic = new Anthropic({ apiKey: ANTHROPIC_API_KEY })
  }
  return _anthropic
}

export function getOpenAI(): OpenAI {
  if (!_openai) {
    _openai = new OpenAI({ apiKey: OPENAI_API_KEY })
  }
  return _openai
}

File: knowledge-mcp/knowledge_scribe/ai.py
"""
Lazy singleton AI clients for Claude and OpenAI.

Import these instead of creating clients at module level in each file.
Clients are created on first access so that importing the module doesn't
fail when API keys are not yet set.
"""

_anthropic_client = None
_openai_client = None


def _get_anthropic():
    global _anthropic_client
    if _anthropic_client is None:
        from anthropic import Anthropic
        _anthropic_client = Anthropic()
    return _anthropic_client


def _get_openai():
    global _openai_client
    if _openai_client is None:
        from openai import OpenAI
        _openai_client = OpenAI()
    return _openai_client


class _LazyClient:
    """Descriptor that lazily initializes an AI client on first attribute access."""

    def __init__(self, factory):
        self._factory = factory

    def __getattr__(self, name):
        client = self._factory()
        return getattr(client, name)


anthropic_client = _LazyClient(_get_anthropic)
openai_client = _LazyClient(_get_openai)
